aten/src/ATen/native/mps/operations/LinearAlgebra.mm

```diff
@@ -560,6 +561,90 @@ bool use_metal_mm(const Tensor& self, const Tensor& other, const Tensor& output)
+      id<MTLCommandBuffer> commandBuffer = mpsStream->commandBuffer();
+
+      MPSMatrixDescriptor* luDescriptor = [MPSMatrixDescriptor matrixDescriptorWithRows:luRows
+                                                                                columns:luCols
+                                                                               matrices:batchSize
+                                                                               rowBytes:luCols * luElemSize
+                                                                            matrixBytes:luRows * luCols * luElemSize
+                                                                               dataType:getMPSDataType(LU)];
+
+      MPSMatrixDescriptor* pivotDescriptor = [MPSMatrixDescriptor matrixDescriptorWithRows:1
+                                                                                   columns:pivotCols
+                                                                                  matrices:batchSize
+                                                                                  rowBytes:sizeof(uint32_t) * luCols
+                                                                               matrixBytes:sizeof(uint32_t) * luCols
+                                                                                  dataType:MPSDataTypeUInt32];
+
+      MPSMatrixDecompositionLU* filter = [[MPSMatrixDecompositionLU alloc] initWithDevice:device
+                                                                                     rows:luRows
+                                                                                  columns:luCols];
+
+      getMPSProfiler().beginProfileKernel(filter, " lu_factor_mps", {LU_});
+
+      for (const auto i : c10::irange(batchSize)) {
+        const uint64_t luBatchOffset = i * luRows * luCols;
+        const uint64_t pivotsBatchOffset = i * pivotCols;
+        MPSMatrix* luMatrix = [[MPSMatrix alloc] initWithBuffer:luBuffer
+                                                         offset:(LU_.storage_offset() + luBatchOffset) * luElemSize
+                                                     descriptor:luDescriptor];
+        MPSMatrix* pivotMatrix =
+            [[MPSMatrix alloc] initWithBuffer:pivotsBuffer
+                                       offset:(pivots.storage_offset() + pivotsBatchOffset) * sizeof(uint32_t)
+                                   descriptor:pivotDescriptor];
+
+        [filter encodeToCommandBuffer:commandBuffer
+                         sourceMatrix:luMatrix
+                         resultMatrix:luMatrix
+                         pivotIndices:pivotMatrix
+                               status:nil];
+      }
+
+      getMPSProfiler().endProfileKernel(filter);
+    }
+  });
+
+  // copy into results
+  LU.copy_(LU_);
+
+  // copy calculated pivots into pivots tensor
+  pivots.copy_(pivots_);
+  // add 1 to pivots to match the 1-based indexing in CPU
+  pivots.add_(1);
```

Comment:
Why is the 1-based indexing needed, is it required by the spec ?  https://pytorch.org/docs/stable/generated/torch.lu.html
The small kernel launches due to copy and add will affect performance...

Reply:
I removed this in the latest commit. I am not quite sure if the pivots parameter that is passed can be expected to be continue, that is why I create `pivots_` and copy after the GPU is done. Maybe this could also be removed. I just wanted to make sure it works the same as the cpu.

Reply:
Makes sense, lets make sure tests are passing...

Reply:
The tests have been updated. IDK why some Linux tests are failing.